js/canvasmanager.ts

```diff
@@ -138,45 +144,46 @@ export class CanvasManager extends StrictEventEmitter<CanvasManagerEvents>() {
       event.preventDefault();
       mouseMoved = false;
       this.drawingImageData = this.ctx.getImageData(0, 0, this.canvas.width, this.canvas.height);
-      this.undoSpliceLists.push(this.tool.onMouseDown(this, point));
+      this.undoStack.push(this.tool.onMouseDown(this, point));
     });
 
     this.canvas.addEventListener("mousemove", event => {
       const point = this.xyFromEvent(event);
       if (point === null) return;
 
       if (this.tool instanceof Eraser) {
-        (this.tool as Eraser).updateIndicator(this.eraserIndicator, point);
+        this.tool.updateIndicator(this.eraserIndicator, point);
       }
 
       if (this.drawingImageData === null) return;
       mouseMoved = true;
-      this.undoSpliceLists[this.undoSpliceLists.length - 1].push(...this.tool!.onMouseMove(this, point));
+      this.undoStack[this.undoStack.length - 1].push(...this.tool!.onMouseMove(this, point));
     });
 
     // document because mouse up outside canvas must also stop drawing
     document.addEventListener("mouseup", event => {
-      const point = this.xyFromEvent(event);
-      if (point === null || this.drawingImageData === null || this.readOnly) return;
+      this.drawingImageData = null;
 
       /*
-      Don't leave empty splice lists to this.undoSpliceLists.
+      Don't leave empty splice lists to this.undoStack.
       Note that they may start off as empty and then become non-empty because of onMouseMove method.
       */
-      if (this.undoSpliceLists.length !== 0 && this.undoSpliceLists[this.undoSpliceLists.length - 1].length === 0) {
-        this.undoSpliceLists.pop();
+      if (this.undoStack.length !== 0 && this.undoStack[this.undoStack.length - 1].length === 0) {
+        this.undoStack.pop();
       }
 
+      const point = this.xyFromEvent(event);
+      if (point === null || this.readOnly) return;
+
       if (!mouseMoved && (this.tool instanceof DrawingTool)) {
         // Draw a dot instead of empty stuff.
         this.objects.pop();
-        const dot = new Circle(point!, this.color, true, 2);
+        const dot = new Circle(point, this.color, true, 2);
         this.objects.push(dot);
         this.draw(dot);
         this.emit("change");
       }
 
-      this.drawingImageData = null;
```

Comment:
Why was this removed?

Reply:
I moved it to the beginning of the method in 5c79fb4